### PR TITLE
Improve .f import

### LINKED ISSUE #21: Support .f files
EDA tools may take a list of command line options from a file. Typically, `-f <options_file>` is used to specify the file, and the options file has a `.f` extension. In their simplest form, `.f` files contain a list of design files, but other command line options may be present as well. 

The goal is to use .f files to create a Sigasi project.

### PR DESCRIPTION
**Pull request by wmeeus**
closes #21

This PR is mostly/only about a .f parser change. We're planning a major overhaul of the project creator for other improvements.

### CODE REVIEW

**Comment by felix-sigasi** [comment]
Am I correct in observing everything in `src/SigasiProjectCreator/DotF` is generated?

**Comment by wmeeus** [comment]
@felix-sigasi most of those files are generated, but `DotFfileParser.py`, `DotF.g4` and `parseFile.py` are not generated.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py [hunk 1]
```diff
@@ -9,12 +9,25 @@
...
 
 from SigasiProjectCreator.ArgsAndFileParser import ArgsAndFileParser
 from .parseFile import parse_dotf
 from ..convertDotFtoCsv import rebase_file
 
 
+def abspath(path):
+    s_path = str(path)
+    if s_path.startswith('\\') or s_path.startswith('/') or s_path[1] == ':' or s_path.startswith('$'):
+        # this is an absolute path in Linux or Windows
```
Comment (felix-sigasi) [inline src/SigasiProjectCreator/DotF/DotFfileParser.py:22]: Please use `os.path.isabs()` instead. What does the `$` mean?

File: src/SigasiProjectCreator/DotF/DotFfileParser.py [hunk 2]
```diff
@@ -9,12 +9,25 @@
...
+        return path
+    return os.path.abspath(path)
+
+
+def expandvars_plus(s):
+    return os.path.expandvars(re.sub(r'\$\((.*)\)', r'${\1}', s))
```
Comment (felix-sigasi) [inline src/SigasiProjectCreator/DotF/DotFfileParser.py:28]: So... `s` may contain parts marked `$(stuff)`, you then transform those into `${stuff}` and then run shell variable expansion? Please document this and why we need this. It seems to me the user should just use `${...}` and let it be replaced by their shell.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py [hunk 3]
```diff
@@ -41,20 +54,13 @@ def __init__(self, filename):
         self.csvfname = str(os.path.splitext(self.dotfname)[0]) + ".csv"
 
         self.filecontent = parse_dotf(filename)
+        parser_expect_library = False
+        parser_expect_dot_f = False
+        newlib = 'work'
         for option in self.filecontent:
+            print(f'*debug* dotf parse: {option}')
```
Comment (felix-sigasi) [inline src/SigasiProjectCreator/DotF/DotFfileParser.py:61]: To be removed when merging, I presume.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py [hunk 4]
```diff
@@ -64,45 +70,74 @@ def __init__(self, filename):
...
-            else:
-                self.library_mapping[file] = library
+        # For now, we'll return a dict of abs path => library
+        # TODO Problem (to be re-solved): files may be mapped to more than one library, in which case this
+        #  approach won't work
+        # Layout moves to ConverterHelper
+
+        # We need to expand environment variables here, before expanding wildcards
+        expanded_path = os.path.expandvars(file)
+        if "*" in expanded_path:
```
Comment (felix-sigasi) [inline src/SigasiProjectCreator/DotF/DotFfileParser.py:110]: Could the `*` not be escaped here?

File: src/SigasiProjectCreator/convertDotFtoCsv.py [hunk 5]
```diff
@@ -12,9 +12,16 @@
 import os
 import csv
 
+def isabspath(path):
+    s_path = str(path)
+    if s_path.startswith('\\') or s_path.startswith('/') or s_path[1] == ':' or s_path.startswith('$'):
+        return True
+    return os.path.isabs(path)
```
Comment (felix-sigasi) [inline src/SigasiProjectCreator/convertDotFtoCsv.py:19]: Duplication, also should just be using os.path.isabs (and expand paths before checking this). The `$` might expand to a relative path.

**Review by felix-sigasi — CHANGES_REQUESTED** [review]
Does the parser system require the generated code to be included in the repository? This makes the codebase needlessly bigger and seemingly more complex, and reviewing harder.

**Comment by felix-sigasi** [comment]
Could you link me to the standardized grammar of this "DotF" language? I'm finding Forth and Forthran but that's what this is parsing.

**Comment by tivervac** [comment]
> Could you link me to the standardized grammar of this "DotF" language? I'm finding Forth and Forthran but that's what this is parsing.

That's the fun part. `.f` has no standard, yet it is used very often in the industry. Sadly it's thus standard-by-implementation

**Comment by felix-sigasi** [comment]
I see. Grammer changes LGTM then.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py (hunk 2 above)
Comment (wmeeus) [reply src/SigasiProjectCreator/DotF/DotFfileParser.py:28]: The `$(...)` syntax occurs in the RiscV repo which we're trying to support.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py (hunk 3 above)
Comment (wmeeus) [reply src/SigasiProjectCreator/DotF/DotFfileParser.py:61]: Sure, done

File: src/SigasiProjectCreator/convertDotFtoCsv.py (hunk 5 above)
Comment (wmeeus) [reply src/SigasiProjectCreator/convertDotFtoCsv.py:19]: This script will be removed in a different PR soon, so duplication will be fixed

File: src/SigasiProjectCreator/DotF/DotFfileParser.py (hunk 1 above)
Comment (wmeeus) [reply src/SigasiProjectCreator/DotF/DotFfileParser.py:22]: We want to be able to detect absolute Windows paths in Linux and vv. A.o. needed for unit tests.

File: src/SigasiProjectCreator/DotF/DotFfileParser.py (hunk 4 above)
Comment (wmeeus) [reply src/SigasiProjectCreator/DotF/DotFfileParser.py:110]: I don't think that an escape is needed here. We test for a wildcard in and handle it if it's present.

**Pull request merged**